[mock_testing, dif_uart] Introduce UART mock tests

silvestrst opened the pull request:
This change introduces a set of UART mock tests to test the public API. The aim is to validate the API implementation in terms of different use-cases.

This PR is related to #1076 and #1224

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+// Copyright lowRISC contributors.
```
I think these files should live side-by-side with the targets they are testing. I don't think there's any danger of confusion with "actual" DIF libraries, and makes finding the tests much easier.

Google C++ does not strictly require this, but it does tacitly recommend it in a few places.

silvestrst replied:
I am happy with `dif_uart_test.cc` and don't really have strong opinion on the naming. One thing I was wondering, whether it is worth calling this directory `mock_testing/mock_tests` instead of `testing`. Just so that anyone who has a look at the directory layout, can see the difference without poking around (we have also sw/device/tests/...). What do you think guys?

alphan replied:
While I don't have a strong feeling on this, I would prefer having tests and their targets in the same directory (as @mcy suggested).

silvestrst replied:
I think I misunderstood what you guys were proposing, do you mean having `dif_uart_test.cc` in `sw/device/lib/dif/` ?

lenary replied:
@silvestrst yes that is what they are proposing. 

I would like it to be quite clear what is production device code and what is not (in this case: host-based C++ tests), and choosing a different directory seems the clearest way of doing so. At lowrisc we seem to agree that `sw/device/tests/dif/dif_uart_test_mocked.cc` is the best place for this test file, and we may move the existing smoke tests in `sw/device/tests` into a subdirectory that makes it clearer what kind of test they are.

alphan replied:
> I think I misunderstood what you guys were proposing, do you mean having `dif_uart_test.cc` in `sw/device/lib/dif/` ?

Yes, that was the proposal but I'm OK with putting them in a separate directory for the sake of clarity and consistency (smoke tests already live in a different directory).

silvestrst replied:
NOTE: still haven't addressed this!

silvestrst replied:
Done.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+
+ private:
+  dif_uart_t dif_uart_;
+  dif_uart_config_t dif_uart_config_;
```
IME, it's more common to just make the member variables protected members and not bother with accessors in test fixtures.

Also, there is also no benefit to having a version that returns a `T&` and aversion that returns a `T*`. They are functionally identical, so it's best to just pick one and stick with it.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+    DifUartTestBasic::SetUp();
+    dif_uart_pointer()->base_addr = dev().region();
+  };
+};
```
I would avoid sticking this initialization in a test fixture. I think tests are easier to read if all of the calls to `dif_uart_` functions (if there are any, at all) happen in the test itself.

alphan replied:
Forgot to mention this, you can also call `_init()` and set the necessary expections instead of modifying the `base_addr` directly. I don't have a strong preference either way.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+   * structure without calling the dif_uart_init() routine.
+   */
+  void SetUp() override {
+    DifUartTestBasic::SetUp();
```
Having to call the base class's definition of the same member function is a code smell, and easy to get wrong.

Disregarding my advice about this class not being the best idea, I think `DifUartTestBasic` should define a pure virtual member function that contains this "extra setup" code, which `SetUp()` then calls.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+ * Difference from InitSuccessDefault, parity=enabled.
+ */
+TEST_F(DifUartTestBasic, InitSuccessParityEven) {
+  dif_uart_config_t *config = &dif_uart_config();
```
This can be `auto` (applies to other similar variable definitions in this file).

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+      'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b',
+      'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'c', 'c',
+      'c', 'c', 'd', 'd', 'd', 'd', 'd', 'd', 'd', 'd',
+  };
```
Consider making this a `constexpr` global of type `std::array` so that you can re-use it in other tests, and so that you can use range-for and size() on it.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+    EXPECT_WRITE32(UART_WDATA_REG_OFFSET, (uint32_t)data[i]);
+  }
+
+  EXPECT_TRUE(dif_uart_bytes_send(&dif_uart(), data, 32, NULL));
```
In C++, prefer `nullptr` over `NULL`.

mcy commented on sw/device/lib/testing/meson.build:
```diff
@@ -76,3 +76,18 @@ test('mock_mmio_test', executable(
+  ],
+  native: true,
+  cpp_args: ['-DMOCK_MMIO'],
+))
```
This target definition should live in the same file as the `_test.cc` file.

silvestrst replied:
NOTE: haven't addressed yet.

silvestrst replied:
We think that the best place for these tests to live is in `sw/device/tests`.

mcy commented on sw/device/lib/testing/meson.build:
```diff
@@ -76,3 +76,18 @@ test('mock_mmio_test', executable(
+    sw_lib_testing_mock_mmio,
+  ],
+  native: true,
+  cpp_args: ['-DMOCK_MMIO'],
```
I suspect you may want to add a `c_args` variant of this, as well. Kind of annoying they're different. =/

silvestrst replied:
NOTE: haven't addressed yet.

silvestrst replied:
Added just in case.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+#include "uart_regs.h"  // Generated.
+}  // extern "C"
+
+namespace dif_uart_test {
```
I'd also take care to include an anonymous namespace here, as well, to prevent name leakage.

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+
+    // Initialise to defaults
+    dif_uart_config_pointer()->parity_enable = kDifUartDisable;
+    dif_uart_config_pointer()->parity = kDifUartParityEven;
```
This configuration should be done in the definition of `dif_uart_config_`, i.e.,
```cc
 private:
  dif_uart_config_t dif_uart_config_{
    /* baudrate = */ 1,
    // ... 
  };
```

Also something something we should maybe consider allowing designated initialization in C++, since it's coming in 20 and GCC/Clang both have working extensions for it...

mcy commented on sw/device/lib/testing/dif/mock_dif_uart_test.cc:
```diff
@@ -0,0 +1,228 @@
+ * parity_type=even, TX,RX FIFOs cleared, and IRQs disabled.
+ */
+TEST_F(DifUartTestBasic, InitSuccessDefault) {
+  EXPECT_WRITE32(UART_CTRL_REG_OFFSET, 0x00010003);
```
Here's an idea: for bitfield-ey registers, how would you feel about mandating
```cc
EXPECT_WRITE32(blah, 0b00'1'010'11 /* etc */);
```
where each separation between apostrophes is a bitfield? I think it would help with test readability.

silvestrst commented:
In the latest push:
* Have introduced additional tests (I think the set of tests is complete now)
* Need to resolve all TODOs (tidy up)
* Possibly simplify

silvestrst commented:
In the latest push:

* Have removed unnecessary test cases (invalid enum value checks)

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+  for (int i = 0; i < 32; ++i) {
+    EXPECT_READ32(UART_STATUS_REG_OFFSET,
+                  0b00000000'00000000'00000000'00000000);
+    EXPECT_WRITE32(UART_WDATA_REG_OFFSET, (uint32_t)bytes_send_data_[i]);
```
Are you sure you want to sign-extend a byte into a word? It might be worthwhile to make `bytes_send_data` contain words rather than bytes.

Also, do not use C-style casts in C++; prefer `static_cast`.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+ * TODO !!!add ...Count values to dif_uart enums!!!
+ */
+TEST_F(DifUartTestUartInterrupt, IrqStateGetSuccess) {
+  for (auto const& irq : uart_irq_reg_map_) {
```
`const auto&`.

silvestrst replied:
Done.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+
+class DifUartTestUartInterrupt : public DifUartTestUartInitialised {
+ protected:
+  std::map<dif_uart_interrupt_t, uint32_t> uart_irq_reg_map_ = {
```
This should also be a constexpr, but the type should be `std::array<std::pair<...>, N>`.

That said, at that point you should forgo `std::pair` entirely and define a separate struct, so that users can write meaningful field names (like `.irq_type` and `.mask`) instead of first and second. In general, avoid pair and tuple unless the STL forces you to use them.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsRestoreSuccess) {
+  uint32_t state = 0b11111111'11111111'11111111'11111111;
```
You should inline this. If you want to refer to all ones or all zeroes I see no reason to not use `-1` or `0`.

silvestrst replied:
I have changed all ones to `0xffffffff` and all zeros to `0`. And I use `constexpr` for `0xffffffff` instead of hardcoding them.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+    EXPECT_TRUE(dif_uart_irq_enable(&dif_uart_, irq.first, kDifUartDisable));
+  }
+
+  EXPECT_EQ(uart_irq_reg_map_.size(), 8); // TODO
```
What does this assertion actually ensure?

silvestrst replied:
This has been deleted. This expectation now lives in `CheckAllInterruptEnumValuesAreCovered`. It just checks whether all of the enum variants are covered - essentially compares `dif_uart.h` num of enum variants with the hardcoded value of how many enum variants are there at the time of writing this test. This way if value is added, it will result in mismatch, and we will know that it hasn't been tested.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+  EXPECT_READ32(UART_FIFO_STATUS_REG_OFFSET,
+                 0b01010101'00000000'01010101'01010101);
+
+  num_bytes = 0xffffffff;
```
I'd avoid reusing out-parameter variables.

silvestrst replied:
Done. Split up in two tests.

mcy commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+    };
+};
+
+class DifUartTestUartInitialised : public DifUartTest {
```
Avoid creating fixtures just for a single behavior. Saving a single line in every test is not worth creating an entire fixture.

One option is to add a member function to DifUartTest along the lines of `InitDevice()` that does what this class's constructor does.

silvestrst commented on sw/device/lib/testing/dif_uart_test.cc:
```diff
@@ -0,0 +1,819 @@
+ * Difference from TxFifoEmptyBytesWrittenSuccess, bytes_written != nullptr.
+ */
+TEST_F(DifUartTestBytesSendReceive, TxFifoEmptyBytesWrittenSuccess) {
+  size_t bytes_written = 0;
```
Done.

silvestrst commented on sw/device/lib/testing/meson.build:
```diff
@@ -28,3 +28,18 @@ test('mock_mmio_test', executable(
+  ],
+  native: true,
+  cpp_args: ['-DMOCK_MMIO'],
+))
```
Added a trailing newline.

silvestrst commented:
In the latest push:
* Attempt to fix a CI build issue.

silvestrst commented:
In the latest push:

Attempt to fix a CI build issue.

silvestrst commented:
In the latest push:
* I have moved the mocked test to live in `sw/device/tests/dif`

silvestrst commented:
Note: I will create a docker container with ubuntu xenial and will try to build. My assumption is to why CI is failing that I have more recent gcc version than in CI, and hence I don't experience the same issue whilst building.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+
+class DifUartSendReceiveTest : public DifUartTest {
+ protected:
+  static const uint8_t kUartFifoSize_ = 32;
```
The "trailing underscore" rule does not apply to static member variables.

Also, this should be a constexpr size_t, since it represents the size of an array in memory. You can downcast it as necessary in tests.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+  /**
+   * Get the underlying uint8_t pointer to data of std::array "send".
+   */
+  const uint8_t *GetSendBytes(void) { return kBytesSendDataArray_.data(); }
```
- `(void)` is a C-ism that should not be used in C++.
- "Simple" accessor-like functions should be in snake_case, not CamelCase.
- Writing accessors for tests is unnecessary, since the member variables of a fixture are available to the tests (hence why they are marked protected). It is very unusual to have these accessors in a fixture.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+      'c', 'c', 'd', 'd', 'd', 'd', 'd', 'd', 'd', 'd',
+  };
+
+  std::array<uint8_t, kUartFifoSize_> kBytesReceiveDataArray_ = {};
```
kConst is only for constants; non-static member variables must be snake_case.

I also think these should both be std::vectors that get initialized to the right capacity in a constructor or something. Using a std::array as a buffer is really eyebrow-raising.

silvestrst replied:
Removed completely.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+  }
+};
+
+class DifUartSendReceiveTest : public DifUartTest {
```
I don't think this fixture provides any value. It does not encapsulate any complicated initialization; tests which need to allocate scratch space should just allocated a `std::vector` and call `reserve` on it; the member functions within should be free functions, since they do not materially depend on class state.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+   * requested, check that remaining bytes of the received array remain unset.
+   */
+  void CompareSendReceiveExpectations(uint8_t num_elements) {
+    assert(num_elements <= kBytesSendDataArray_.size());
```
Do not use C `assert`, use `EXPECT_LE` or `ASSERT_LE`, which work correctly with gtest.

silvestrst replied:
Removed completely.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+  void CompareSendReceiveExpectations(uint8_t num_elements) {
+    assert(num_elements <= kBytesSendDataArray_.size());
+
+    for (uint8_t i = 0; i < num_elements; ++i) {
```
Use `int` as your iteration variable type unless you can't avoid it. This applies to the entire file.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+    // Check that nothing that hasn't been requested, is written.
+    for (uint8_t i = num_elements; i < kBytesSendDataArray_.size(); ++i) {
+      EXPECT_EQ(kBytesReceiveDataArray_[i], 0);
+    }
```
C++ implements equality for various containers. Consider using std::vector, which will allow you to compress these checks into a single EXPECT_EQ. I also suspect there is an EXPECT_THAT matcher that can help you.
https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+      1048576,            /* clk_freq_hz */
+      kDifUartDisable,    /* parity_enable */
+      kDifUartParityEven, /* parity */
+  };
```
Member variables go after member functions, as a general rule.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+typedef struct DifUartInterruptPair {
+  dif_uart_interrupt_t name;
+  uint32_t value;
+} DifUartInterruptPairStruct;
```
Do not use `typedef` for anything in C++. If you need an alias, use `using`. If you need to declare a struct, declare it as
```c
struct InterruptPair {
  // ...
};
```

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+#include "uart_regs.h"  // Generated.
+}  // extern "C"
+
+namespace {
```
Please wrap this in an external, named namespace, to avoid ADL silliness due to all these names being in global scope.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+}  // extern "C"
+
+namespace {
+constexpr uint32_t kUint32Max = 0xffffffff;
```
If you need the maximum of some integral type, prefer `std::numeric_limits<T>::max()`.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+
+namespace {
+constexpr uint32_t kUint32Max = 0xffffffff;
+constexpr uint8_t kDifUartInterruptCount = kDifUartInterruptLast + 1;
```
Is there any reason this needs to be byte-sized and can't be decalred as a size_t?

silvestrst replied:
Made it size_t.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+   * dif_uart_init() must always result in sequence of these calls before any
+   * other initialisation/configuration register access.
+   */
+  void DeviceResetExpectations(void) {
```
If you want to wrap a bunch of EXPECTs, consider a name like `ExpectReset()`.

silvestrst replied:
It is still pretty easy to determine where the fault is, as it will tell you which test failed. So I will use expects in functions for now, and maybe I will change my mind later.

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+// Copyright lowRISC contributors.
```
Google C++ requires that this file live adjacent to the header/TU pair it tests. Please move this file into `sw/device/lib/dif`.

silvestrst replied:
We had this discussion earlier with @lenary and @alphan included. @lenary and myself agree that it is weird to keep CI only code together with the production code in lib.

mcy replied:
I strongly disagree that that is weird. Tests are not CI-only code; they act as a kind of documentation that should be easy to find for anyone using those DIFs, and, as such, should live alongside the code that they test.

silvestrst replied:
Well I think more to the point that something that will never be compiled and ran for the target lives together with something that always is compiled and is ran on the target is a bit confusing.

mcy replied:
I chatted with Garret about this. Normally, I would still push to have everything in the same directory, but the real point is that it needs to be easy to find unit tests _and_ integration tests, and those can't all live in the same directory, so we're going to need a more disciplined story.

I think that, for now, you should just do whatever, or follow the existing convention of the two existing in-tree unit tests. Placement of tests is something we need to decide and document, rather than just doing whatever.

Also, me and Garret decided that the "_test" prefix is potentially inappropriate, given that it's a very overloaded term, so we might want to go with something like "_unittest" (the fact that the tests use mocks is pretty irrelevant and should not be in the name: print_test.cc, for example, does not use mocks, but is also a unit test).

lenary replied:
I'm happy for the file suffix to be `_unittest` (rather than `_test_mocked`).

mcy commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,734 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cassert>
```
Standard library includes should go between the header-under-test include and the other project includes.

Also, do not use C assertions.

silvestrst replied:
Done.

silvestrst commented:
@mcy thanks again for your review, much appreciated. Does this change look better now?

lenary commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,707 @@
+};
+
+/**
+ * Sets init reset expectations
```
We may need to abandon pipe characters: Doxygen doesn't support using them to denote monospaced text (it does support markdown's backticks though). This is an issue to address later.

mcy replied:
Sure, but we can do a global find-and-replace later. Things that intend to be monospaced should be denoted *somehow*.

lenary commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,707 @@
+ * Every sent byte by the "send bytes" routine is expected to result in the
+ * STATUS read of 0 (FIFO not full), and write to WDATA.
+ */
+void ExpectUartSendBytes(mock_mmio::MockDevice &dev, int num_elements) {
```
Presumably we cannot use `const mock_mmio::MockDevice &dev` here because the mock mutates interior state?

mcy replied:
Yeah, because .region() isn't marked as const, because it casts `this` to a `void*`, which later gets non-const functions called on it.

silvestrst replied:
Done.

silvestrst commented:
Outstanding issues:
* @mcy latest review comments should be addressed.
* This PR should be refactored to use new `EXPECT_*` macros from #1647 .

silvestrst commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,707 @@
+}
+
+TEST_F(DifUartTest, InitSuccessParityEven) {
+  // parity=enabled, parity_type=even.
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,707 @@
+}
+
+TEST_F(DifUartTest, InitSuccessDefault) {
+  ExpectReset(this->dev());
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test_mocked.cc:
```diff
@@ -0,0 +1,707 @@
+                                     receive_bytes.data(), &bytes_read));
+
+  EXPECT_EQ(bytes_read, 0);
+  EXPECT_TRUE(std::all_of(receive_bytes.begin(), receive_bytes.end(),
```
Done.

silvestrst commented:
I think any further refactoring only makes sense after #1647 has landed.

silvestrst commented:
In the latest push:
* Some refactoring to match @mcy SPI unittest where possible.
* Small `dif_uart` refactoring (see the other two commits).

By the looks of it in some places `clang-format ` has done some weird formatting, not sure how I feel about it.

alphan commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+  // Get the last IRQ state.
+  EXPECT_READ32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_RX_PARITY_ERR, false}});
```
Is the intention to return 0 here?

silvestrst replied:
Good catch, not anymore I think its from previous iterations of the change. Thanks.

alphan commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+                                   kDifUartEnable));
+}
+
+TEST_F(IrqEnableTest, Success) {
```
There are 9 options but this test exercises only 2 of them. IMHO, remaining 7 also deserve to be tested :)

silvestrst replied:
Initially I was testing all of them, although when I came to my PLIC unittests it proved to be problematic and there is no good way to test every bit of multireg type of IE, IP, LE registers without making the tests more complicated.

There are trade-offs with every approach. The reason why I went with simply testing the last and the first variant because the API follows the same logic for all of the variants. Yes, there is more chance of missing a bug, but I think it is a reasonable compromise for now, and we can come back to it later if need to.

silvestrst replied:
That is exactly the reason why I would love to see things like `UART_INTR_STATE_RX_WATERMARK` as an enum, so that it is possible to simply loop through it. And it should be easily doable for peripherals like UART (the IRQs are internal information to the peripheral), it is much trickier with peripherals like PLIC, as a lot of information comes from `top_earlgrey`.

alphan commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  EXPECT_FALSE(dif_uart_irq_force(nullptr, kDifUartInterruptTxWatermark));
+}
+
+TEST_F(IrqForceTest, Success) {
```
Same here.

silvestrst replied:
Same as above.

alphan commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class RxBytesAvailableTest : public UartTest {};
+
+TEST_F(RxBytesAvailableTest, UartNull) {
+  size_t num_bytes;
+  EXPECT_FALSE(dif_uart_rx_bytes_available(nullptr, &num_bytes));
+}
+
+TEST_F(RxBytesAvailableTest, BytesNull) {
+  EXPECT_FALSE(dif_uart_rx_bytes_available(&dif_uart_, nullptr));
+}
```
Please see my previous comments on null checks.

silvestrst replied:
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class InitTest : public UartTest {};
+
+TEST_F(InitTest, UartNull) {
+  EXPECT_FALSE(dif_uart_init(base_addr_, &dif_uart_config_, nullptr));
+}
+
+TEST_F(InitTest, ConfigNull) {
+  EXPECT_FALSE(dif_uart_init(base_addr_, nullptr, &dif_uart_));
+}
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class ConfigTest : public UartTest {};
+
+TEST_F(ConfigTest, UartNull) {
+  EXPECT_FALSE(dif_uart_init(base_addr_, &dif_uart_config_, nullptr));
+}
+
+TEST_F(ConfigTest, ConfigNull) {
+  EXPECT_FALSE(dif_uart_init(base_addr_, nullptr, &dif_uart_));
+}
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  EXPECT_TRUE(dif_uart_configure(&dif_uart_, &dif_uart_config_));
+}
+
+class WatermarkRxSetTest : public UartTest {};
```
I see your point of view, but I personally prefer, where possible, having per function tests. I think I will leave this one as it is.

mcy replied:
I think I did something similar to Silvestrs in one of mine; I think this test layout is fine.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  }
+};
+
+TEST_F(BytesSendTest, UartNull) {
+  EXPECT_FALSE(dif_uart_bytes_send(nullptr, kBytesArray.data(), 1, nullptr));
+}
+
+TEST_F(BytesSendTest, DataNull) {
+  EXPECT_FALSE(dif_uart_bytes_send(&dif_uart_, nullptr, 1, nullptr));
+}
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXFULL, false}});
+
+  // Set expectations for the byte to be set
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXFULL, false}});
```
Well, unfortunately this is required as a polled read/write is done in terms of a static `BytesSend` function that is common between polled and non-polled public functions. I guess I could look into splitting that check out of the static function, but I think it is a subject for a different change.

I actually was thinking about it myself yesterday. Thank you for a thorough review :)

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class BytesReceivePolledTest : public UartTest {};
+
+TEST_F(BytesReceivePolledTest, UartNull) {
+  uint8_t byte;
+  EXPECT_FALSE(dif_uart_byte_receive_polled(nullptr, &byte));
+}
+
+TEST_F(BytesReceivePolledTest, ByteNull) {
+  EXPECT_FALSE(dif_uart_byte_receive_polled(&dif_uart_, nullptr));
+}
```
Same as above.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_RXEMPTY, false}});
+
+  // Set expectations for the byte to be read
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_RXEMPTY, false}});
```
Same as above.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class IrqStateGetTest : public UartTest {};
+
+TEST_F(IrqStateGetTest, UartNull) {
+  dif_uart_enable_t state;
+  EXPECT_FALSE(
+      dif_uart_irq_state_get(nullptr, kDifUartInterruptTxWatermark, &state));
+}
+
+TEST_F(IrqStateGetTest, StateNull) {
+  EXPECT_FALSE(dif_uart_irq_state_get(&dif_uart_, kDifUartInterruptTxWatermark,
+                                      nullptr));
+}
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+
+class UartTest : public Test, public MmioTest {
+ protected:
+  void DeviceReset() {
```
Done.

silvestrst commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,657 @@
+  }
+};
+
+TEST_F(BytesReceiveTest, UartNull) {
+  std::vector<uint8_t> receive_bytes(1);
+  EXPECT_FALSE(
+      dif_uart_bytes_receive(nullptr, 1, receive_bytes.data(), nullptr));
+}
+
+TEST_F(BytesReceiveTest, DataNull) {
+  EXPECT_FALSE(dif_uart_bytes_receive(&dif_uart_, 1, nullptr, nullptr));
+}
```
Done.

silvestrst commented:
@alphan thank you for your review, your comments were very useful.

In the lastest push:
* Have addressed most of @alphan comments.
* Rebased on top of master.

mcy commented on sw/device/lib/dif/dif_uart.h:
```diff
@@ -10,6 +10,8 @@
 
 #include "sw/device/lib/base/mmio.h"
 
+#define DIF_UART_FIFO_SIZE_BYTES 32u
```
Do you think you could replace this with a `const`? if not, can you at least add a Doxygen comment?

silvestrst replied:
Good point, this is not used anywhere in the header and can be a const. I will add a comment as well.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,644 @@
+   */
+  void ExpectSendBytes(int num_elements) {
+    for (int i = 0; i < num_elements; ++i) {
+      uint32_t value = static_cast<uint32_t>(kBytesArray[i]);
```
Maybe add a check to make sure `i` doesn't go past the end of `kBytesArray`?

silvestrst replied:
Good point, will do.

silvestrst replied:
Done.

silvestrst commented:
I think this change is ready for a merge.

In the latest push:
* Addressed all @mcy comments.
* Rebased with master.
* `namespace dif_uart_test_mocked` -> `namespace dif_uart_test`

lenary commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,645 @@
+class WatermarkTxSetTest : public UartTest {};
+
+TEST_F(WatermarkTxSetTest, UartNull) {
+  EXPECT_FALSE(dif_uart_watermark_tx_set(nullptr, kDifUartWatermarkByte1));
```
You should have a test here for the error condition associated with `kDifUartWatermarkByte30`. Here I wrote it for you:
```c
  EXPECT_FALSE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte30));
```

silvestrst replied:
Good point, I will add it.

lenary commented on sw/device/tests/dif/dif_uart_test.cc:
```diff
@@ -0,0 +1,645 @@
+
+  dif_uart_enable_t last_state = kDifUartEnable;
+  EXPECT_TRUE(
+      dif_uart_irq_state_get(&dif_uart_, kDifUartInterruptLast, &last_state));
```
Do you think it would be more clear if you used `kDifUartInterruptRxParityErr` here?

Reasons to use `…Last`: it ensures we have full interrupt coverage (are there other ways of doing this?).
Reasons not to use `…Last`: it turns these enum constants into parts of the public API, which I'm not sure we intended when we added them (they were for internal use during iteration within the DIF).

Reasons to use `…RxParityErr`: it corresponds to the interrupt register field named in `EXPECT_READ32`.

I'll note you didn't use `…First` at the top of this test.

silvestrst replied:
Sure, ill do that.

lenary commented on sw/device/tests/dif/meson.build:
```diff
@@ -17,3 +17,19 @@ test('dif_spi_device_test', executable(
+  sources: [
+    'dif_uart_test.cc',
+    meson.source_root() / 'sw/device/lib/dif/dif_uart.c',
+    [hw_ip_uart_reg_h],
```
This does not need to be in `[]`.
```suggestion
    hw_ip_uart_reg_h,
```

silvestrst replied:
Will do.

silvestrst commented:
In the latest push:
* Addressed @lenary review comments.
